### PR TITLE
Initial commit of completed optimal extraction notebook.

### PR DESCRIPTION
**Pull request by gkanarek**
This notebook should be ready for review.

### CODE REVIEW

**Comment by dtn5ah** [comment]
Hi @gkanarek, thanks for the PR! I'm currently taking a look and I'm having issues installing; was wondering if you could help me try and resolve those?

Firstly, for some reason, installing the requirements file `pip install -r requirements.txt` couldn't get past the JWST package, complaining about needing numpy before it can be installed (which is the same reason why the built_and_check is failing). So I installed numpy first, before reinstalling the requirements.txt. However, afterwards, the JWST package gave the same error as listed: https://github.com/spacetelescope/jwst/issues/3164. By just listing 'JWST' in the requirements file, it pulls the version off of pip, which is v0.0.0. I presume you're using a more modern version of the JWST pipeline. So the requirements file should instead point to github, using whichever version of the JWST pipeline you're using. After a quick google search, you can change the `jwst` line in your requirements file to `git+git://github.com/spacetelescope/jwst@0.X.X#egg=jwst` (As an example, for JWST version v0.16.2, `git+git://github.com/spacetelescope/jwst@0.16.2#egg=jwst` )

I tried installing the latest stable release (0.16.2) but ran into the following error. That being said, I don't want to try and dig too deep before I know which version you're intending this notebook to use:
```
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
 in 
----> 1 data_model = ImageModel(s2d_file)
      2 resampled_2d_image = data_model.data # if multiple SCI extensions, also specify EXTVER
      3 weights_2d_image = data_model.wht # we will use this to estimate the per-pixel variance later
      4 
      5 image_shape = resampled_2d_image.shape
...
f:\STScI\gitRepos\dat_pyinthesky\jdat_notebooks\optimal_extraction\env\lib\site-packages\asdf\generic_io.py in get_file(init, mode, uri, close)
   1191                 fd = atomicfile.atomic_open(realpath, realmode)
   1192             else:
-> 1193                 fd = open(realpath, realmode)
   1194             fd = fd.__enter__()
   1195             return RealFile(fd, mode, close=True, uri=uri)

FileNotFoundError: [Errno 2] No such file or directory: 'F:\\STScI\\gitRepos\\dat_pyinthesky\\jdat_notebooks\\optimal_extraction\\env\\lib\\site-packages\\asdf\\schemas\\stsci.edu\\jwst_datamodel\\image.schema.yaml'
```
Thank you!
Duy

**Comment by gkanarek** [comment]
I can definitely update the requirements.txt line for jwst - I was using version 0.16.1. I'm very confused by that file path -- in my install, the schemas are stored in the `jwst.datamodels` package, not in `asdf`. Is that something that changed from 0.16.1 to 0.16.2??

**Comment by dtn5ah** [comment]
I'm not sure, I was going to dig into this specific issue once I nailed down your `jwst` version just to eliminate any possible variables before I go down a rabbit hole. With that, I can try and pin down what the issue is; thanks!

File: jdat_notebooks/optimal_extraction/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+numpy
+astropy
+matplotlib
+scipy
+ipywidgets
+jwst
```
Comment (dtn5ah) [inline jdat_notebooks/optimal_extraction/requirements.txt:6]: As @gkanarek and I discussed, `jwst` pulls v0.0.0 from PyPI. @gkanarek is using jwst version v0.16.1. This is required, as v0.16.2 appears to break this on my end. 
```suggestion
git+git://github.com/spacetelescope/jwst@0.16.1#egg=jwst
```

File: jdat_notebooks/optimal_extraction/Spectral Extraction.ipynb [hunk 2]
```diff
@@ -0,0 +1,1746 @@
...
+    "See the WebbPSF documentation for a full list of instrument settings."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "instrument = NIRSpec()\n",
```
Comment (dtn5ah) [inline jdat_notebooks/optimal_extraction/Spectral Extraction.ipynb:1410]: I'm still getting the same error as before: https://github.com/spacetelescope/dat_pyinthesky/pull/105#pullrequestreview-439483053

**Pull request merged**